File: hot100/04_Stack/0155_Min_Stack/solution.go
package main

import "fmt"

const (
	grn = "\033[32m"
	red = "\033[31m"
	bld = "\033[1m"
	rst = "\033[0m"
)

// ─── Solution ────────────────────────────────────────────────────────────────

type MinStack struct {
	// TODO: implement
}

func Constructor() MinStack {
	// TODO: implement
	return MinStack{}
}

func (ms *MinStack) Push(val int) {
	// TODO: implement
}

func (ms *MinStack) Pop() {
	// TODO: implement
}

func (ms *MinStack) Top() int {
	// TODO: implement
	return 0
}

func (ms *MinStack) GetMin() int {
	// TODO: implement
	return 0
}

// ─── Test Harness ─────────────────────────────────────────────────────────────

func main() {
	passed, total := 0, 0

	checkInt := func(got, want int, label string) {
		total++
		if got == want {
			fmt.Printf(grn+"✓"+rst+"  %-40s →  %d\n", label, got)
			passed++
		} else {
			fmt.Printf(red+"✗"+rst+"  "+bld+"%-40s"+rst+"    want=%d  got=%d\n", label, want, got)
		}
	}

	// Example 1: push(-2), push(0), push(-3), getMin, pop, top, getMin
	ms1 := Constructor()
	ms1.Push(-2)
	ms1.Push(0)
	ms1.Push(-3)
	checkInt(ms1.GetMin(), -3, "ex1: getMin after push(-2,0,-3)")
	ms1.Pop()
	checkInt(ms1.Top(),    0,  "ex1: top after pop")
	checkInt(ms1.GetMin(), -2, "ex1: getMin after pop")

	// Edge case: single element
	ms2 := Constructor()
	ms2.Push(5)
	checkInt(ms2.Top(),    5, "edge: single top")
	checkInt(ms2.GetMin(), 5, "edge: single getMin")

	// Edge case: duplicate minimums
	ms3 := Constructor()
	ms3.Push(1)
	ms3.Push(1)
	checkInt(ms3.GetMin(), 1, "edge: dup min before pop")
	ms3.Pop()
	checkInt(ms3.GetMin(), 1, "edge: dup min after pop")

	// Edge case: ascending then pop
	ms4 := Constructor()
	ms4.Push(3)
	ms4.Push(5)
	ms4.Push(7)
	checkInt(ms4.GetMin(), 3, "edge: asc getMin = 3")
	ms4.Pop()
	ms4.Pop()
	checkInt(ms4.Top(), 3, "edge: asc top after 2 pops = 3")

	fmt.Printf("\n%s\n", "--------------------------------------------")
	if passed == total {
		fmt.Printf(grn+"%d / %d passed\n"+rst, passed, total)
	} else {
		fmt.Printf(red+"%d / %d passed\n"+rst, passed, total)
	}
}

File: hot100/04_Stack/0155_Min_Stack/solution.cpp
#include <bits/stdc++.h>
using namespace std;

#define GRN "\033[32m"
#define RED "\033[31m"
#define BLD "\033[1m"
#define RST "\033[0m"

// ─── Solution ────────────────────────────────────────────────────────────────

class MinStack {
private:
    stack<int> s;  // main stack
    stack<int> m;  // min stack  NON-INC order

public:
    MinStack() {

    }

    void push(int val) {
        s.push(val);
        if (m.empty() or m.top() >= val) 
            m.push(val);
    }

    void pop() {
        if (m.top() == s.top()) 
            m.pop();
        s.pop();
    }

    int top() {
        return s.top();
    }

    int getMin() {
        return m.top();
    }
};

// ─── Test Harness ─────────────────────────────────────────────────────────────

int main() {
    int passed = 0, total = 0;

    auto checkInt = [&](int got, int want, const string& label) {
        total++;
        if (got == want) {
            printf(GRN "✓" RST "  %-40s →  %d\n", label.c_str(), got);
            passed++;
        } else {
            printf(RED "✗" RST "  " BLD "%-40s" RST "    want=%d  got=%d\n",
                   label.c_str(), want, got);
        }
    };

    // Example 1: push(-2), push(0), push(-3), getMin, pop, top, getMin
    {
        MinStack ms;
        ms.push(-2);
        ms.push(0);
        ms.push(-3);
        checkInt(ms.getMin(), -3, "ex1: getMin after push(-2,0,-3)");
        ms.pop();
        checkInt(ms.top(),    0,  "ex1: top after pop");
        checkInt(ms.getMin(), -2, "ex1: getMin after pop");
    }

    // Edge case: single element
    {
        MinStack ms;
        ms.push(5);
        checkInt(ms.top(),    5, "edge: single top");
        checkInt(ms.getMin(), 5, "edge: single getMin");
    }

    // Edge case: duplicate minimums
    {
        MinStack ms;
        ms.push(1);
        ms.push(1);
        checkInt(ms.getMin(), 1, "edge: dup min before pop");
        ms.pop();
        checkInt(ms.getMin(), 1, "edge: dup min after pop");
    }

    // Edge case: ascending then pop reveals new min
    {
        MinStack ms;
        ms.push(3);
        ms.push(5);
        ms.push(7);
        checkInt(ms.getMin(), 3, "edge: asc getMin = 3");
        ms.pop();
        ms.pop();
        checkInt(ms.top(),    3, "edge: asc top after 2 pops = 3");
    }

    printf("\n%s\n", string(44, '-').c_str());
    if (passed == total)
        printf(GRN "%d / %d passed\n" RST, passed, total);
    else
        printf(RED "%d / %d passed\n" RST, passed, total);
    return (passed == total) ? 0 : 1;
}
